py/picca/fitter2/control_mpi.py

```diff
@@ -0,0 +1,332 @@
+
+        # Run 2D grid
+        elif dim == 2:
+            par1 = list(self.chi2.dic_chi2scan.keys())[0]
+            par2 = list(self.chi2.dic_chi2scan.keys())[1]
+            values1 = self.chi2.dic_chi2scan[par1]['grid']
+            values2 = self.chi2.dic_chi2scan[par2]['grid']
+            grid = []
+            for v1 in values1:
+                for v2 in values2:
+                    grid += [[v1, v2]]
+            grid = np.asarray(grid)
+            num_runs = len(grid)  # Number of points to be run
+
+            if num_runs <= num_cpus:
+                result = mpi_run_2d_set(num_runs, grid, par1, par2)
+            else:
+                j = 0
+                for __ in range(num_runs // num_cpus):
+                    result += [mpi_run_2d_set(num_cpus, grid[j:j+num_cpus], par1, par2)]
+                    j += num_cpus
+                rest = num_runs % num_cpus
+                if rest != 0:
+                    result += [mpi_run_2d_set(rest, grid[-rest:], par1, par2)]
+
+        # If we are on CPU #0 concatenate results to match the normal output
+        if cpu_rank == 0 and num_runs > num_cpus:
+            temp = result[0]
+            for res in result[1:]:
+                temp = np.r_[temp, res]
+            result = temp
+
+        self.chi2.dic_chi2scan_result = {}
+        self.chi2.dic_chi2scan_result['params'] = np.asarray(np.append(sorted(self.chi2.best_fit.values), ['fval']))
+        self.chi2.dic_chi2scan_result['values'] = np.asarray(result)
+
+        # Set all parameters to where they were before
+        for d in self.chi2.data:
+            store_d_pars_init = store_data_pars[d.name]['init']
+            store_d_par_error = store_data_pars[d.name]['error']
+            store_d_par_fixed = store_data_pars[d.name]['fixed']
+            for name in d.pars_init.keys():
+                d.pars_init[name] = store_d_pars_init[name]
+            for name in d.par_error.keys():
+                d.par_error[name] = store_d_par_error[name]
+            for name in d.par_fixed.keys():
+                d.par_fixed[name] = store_d_par_fixed[name]
+
+    def _mpi_fastMC(self):
+        '''
+        Run fastMC mocks in parallel using MPI.
+        Each CPU will output its own .h5 file when it finishes its allotted computation.
+
+        In some cases the model fails and we get ValueError when minimizing. For these
+        runs, the output will be filled with inf and should be ignored in the analysis until
+        we figure out why they happen.
+        '''
+        if not hasattr(self.chi2, "nfast_mc"):
+            return
+
+        mpi_comm = MPI.COMM_WORLD
+        cpu_rank = mpi_comm.Get_rank()
+        num_cpus = mpi_comm.Get_size()
+
+        # Seed is incremented from the input value by the CPU #
+        np.random.seed(self.chi2.seedfast_mc + cpu_rank)
+        nfast_mc = self.chi2.nfast_mc // num_cpus
+        if self.chi2.nfast_mc % num_cpus != 0:
+            nfast_mc += 1
```

Comment:
Why do we do this here? Please document.

Reply:
How do you know that adding one set is enough / better? Is the goal to make sure that all CPUs run exactly the same number of cases? If so, shouldn't the number of extra cases to add depend on the specific case?

Reply:
So let me give an example, which I hope is clearer. Say we have 5 cores and we ask for 13 mocks. In this case, there are three options of what we can do:
1. Compute 3 * 5 = 15 mocks as the smallest multiple larger than the number asked.
2. Compute 2 * 5 = 10 mocks as the largest multiple smaller than the number asked.
3. Compute 13 mocks, but for the last set 2 CPUs are idle.

Number 3 is much harder to implement, and wastes computing time. I think number 2 is not ideal because the user wants 13 mocks, so why compute less than asked? Which leaves number one as the best option that I see. And this is the one I did.

Of course, ideally, this shouldn't happen and the number of mocks should be a multiple of the number of CPUs. But I don't want to stop the run with an error if it isn't as it may have queued for a long time to start running.

Reply:
I agree with should do option (1), I just didn't understand the code. 

I got it now, I didn't realyze the "nfast_mc" was actually "nfast_mc_per_processor". I thought it was "updated_nfast_mc". 

Everything makes sense now, sorry for the noise.